pkg/operation/botanist/component/deploy.go

```diff
@@ -16,60 +16,78 @@ package component
 
 import "context"
 
-// OpDestroy creates a DeployWaiter which calls Destroy instead of create
-// and WaitCleanup instead of Wait
-func OpDestroy(dw ...DeployWaiter) DeployWaiter {
+// OpDestroy creates a Deployer which calls Destroy instead of Deploy.
+func OpDestroy(d ...Deployer) Deployer {
 	return &deploy{
-		dw:          dw,
-		invert:      true,
-		destroyOnly: false,
+		deployers: d,
+		invert:    true,
+		wait:      false,
 	}
 }
 
-// OpDestroyAndWait creates a Deployer which calls Destroy instead of create
-// and waits for destruction.
-func OpDestroyAndWait(dw ...DeployWaiter) Deployer {
+// OpWait creates a DeployWaiter which calls Wait .
+func OpWait(dw ...DeployWaiter) DeployWaiter {
 	return &deploy{
-		dw:          dw,
-		invert:      true,
-		destroyOnly: true,
+		deployWaiters: dw,
+		invert:        false,
+		wait:          true,
 	}
 }
 
-// OpWaiter creates a Deployer which calls waits for each operation.
-func OpWaiter(dw ...DeployWaiter) Deployer {
+// OpDestroyAndWait creates a DeployWaiter which calls Destroy instead of Deploy, and WaitCleanup.
+func OpDestroyAndWait(dw ...DeployWaiter) DeployWaiter {
 	return &deploy{
-		dw:          dw,
-		invert:      false,
-		destroyOnly: true,
+		deployWaiters: dw,
+		invert:        true,
+		wait:          true,
 	}
 }
 
-// NoOp does nothing
+// OpDestroyWithoutWait creates a DeployWaiter which calls Destroy instead of Deploy.
+func OpDestroyWithoutWait(dw ...DeployWaiter) DeployWaiter {
+	return &deploy{
+		deployWaiters: dw,
+		invert:        true,
+		wait:          false,
+	}
+}
+
+// NoOp does nothing.
 func NoOp() DeployWaiter { return &deploy{} }
 
 type deploy struct {
-	invert      bool
-	destroyOnly bool
-	dw          []DeployWaiter
+	deployers     []Deployer
+	deployWaiters []DeployWaiter
```

Comment:
You could simply store a single slice of `Deployers` and then in `Deploy`, `Wait`, and friends check if the deployer implements `Waiter`, e.g.:

```go
		if err := deployer.Deploy(ctx); err != nil {
			return err
		}
		
		if waiter, ok := deployer.(Waiter); d.wait && ok {
			if err := waiter.Wait(ctx); err != nil {
				return err
			}
		}
```

Reply:
Right, I was doing it this way at some point in time, but due to back-and-forth dropped it again. Definitely makes sense. PTAL :)